### PR TITLE
fix gradient norm tracking for row_log_interval > 1

### LINKED ISSUE #3487: Gradient norms are not logged unless row_log_interval==1
## 🐛 Bug
In version 0.9 the guards to calculate the gradient norms and then log the metrics can't be satisfied in the same batch unless the row_log_interval  is 1. In most places the guard seems to be `(batch_idx + 1) % self.row_log_interval == 0` such as here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/b40de5464a953ff5866a255f4670d318bd8fd65a/pytorch_lightning/trainer/training_loop.py#L749-L757

However in `run_batch_backward_pass` it is `batch_idx % self.row_log_interval == 0`

https://github.com/PyTorchLightning/pytorch-lightning/blob/b40de5464a953ff5866a255f4670d318bd8fd65a/pytorch_lightning/trainer/training_loop.py#L929-L939

### To Reproduce

1. Run the code sample below (taken from #1527 ).
2. Confirm that gradients are not being logged in tensorboard.
3. Change row_log_interval to 1 and rerun the code.
4 Confirm that gradients are now being logged.

#### Code sample

```
import pytorch_lightning as pl
from torch.utils.data import TensorDataset, DataLoader
from pytorch_lightning.loggers.tensorboard import TensorBoardLogger
from torch.optim import SGD
import torch.nn as nn
import torch


class MWENet(pl.LightningModule):
    def __init__(self):
        super(MWENet, self).__init__()

        self.first = nn.Conv2d(1, 1, 3)
        self.second = nn.Conv2d(1, 1, 3)
        self.loss = nn.L1Loss()

    def train_dataloader(self):
        xs, ys = torch.zeros(16, 1, 10, 10), torch.ones(16, 1, 6, 6)
        ds = TensorDataset(xs, ys)
        return DataLoader(ds)

    def forward(self, xs):
        out = self.first(xs)
        out = self.second(out)
        return out

    def configure_optimizers(self):
        first = SGD(self.first.parameters(), lr=0.01)
        second = SGD(self.second.parameters(), lr=0.01)
        return [second, first]

    def training_step(self, batch, batch_idx, optimizer_idx):
        xs, ys = batch
        out = self.forward(xs)
        return {'loss': self.loss(out, ys)}


net = MWENet()
logger = TensorBoardLogger('tb_logs', name='testing')
trainer = pl.Trainer(
    track_grad_norm=2,
    row_log_interval=2,
    max_epochs=50,
    logger=logger)
trainer.fit(net)

```
### Expected behavior

Gradients should be logged if track_grad_norm is True

### Environment

* CUDA:
        - GPU:
                - GeForce GTX 1080 Ti
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.19.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.6.0
        - pytorch-lightning: 0.9.0
        - tensorboard:       2.2.1
        - tqdm:              4.48.2
* System:
        - OS:                Windows
        - architecture:
                - 64bit
                - WindowsPE
        - processor:         AMD64 Family 23 Model 113 Stepping 0, AuthenticAMD
        - python:            3.7.9
        - version:           10.0.19041

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?
Fixes  #3487

**Test fails on master, replicating the bug report**

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@rohitgr7 The test I wrote replicates the bug but it is not very elegant. I'll try to make it better. If you have any suggestions let me know.

File: tests/models/test_grad_norm.py [hunk 1]
```diff
@@ -73,3 +74,23 @@ def test_grad_tracking(tmpdir, norm_type, rtol=5e-3):
...
 
         assert np.allclose(log, mod, rtol=rtol)
+
+
+@pytest.mark.parametrize("row_log_interval", [1, 2, 3])
+def test_grad_tracking_interval(tmpdir, row_log_interval):
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        track_grad_norm=2,
+        row_log_interval=2,
```
Comment (Tim-Chard) [inline tests/models/test_grad_norm.py:84]: Shouldn't this be `row_log_interval=row_log_interval`

File: tests/models/test_grad_norm.py [hunk 2]
```diff
@@ -73,3 +74,23 @@ def test_grad_tracking(tmpdir, norm_type, rtol=5e-3):
...
+        max_steps=10,
+    )
+
+    with patch.object(trainer.logger, "log_metrics") as mocked:
+        model = EvalModelTemplate()
+        trainer.fit(model)
+        expected = trainer.global_step // 2
```
Comment (Tim-Chard) [inline tests/models/test_grad_norm.py:91]: and here should be `expected = trainer.global_step // row_log_interval`

**Review by Tim-Chard — COMMENTED** [review]
You beat me to it! I was going to change the test above to include the `row_log_interval` parameter to something like the code snippet below but it doesn't work because of the corner cases (the logging of the first and last batch). You could handle them but then it seems like that would make the test just as complex as the code being tested so I think yours is the way to go. 

In case you can think of some simple way to do it here is the partially modified test
```
@pytest.mark.parametrize("norm_type", [1., 1.25, 2, 3, 5, 10, 'inf'])
@pytest.mark.parametrize("log_interval", [1, 2, 20])
def test_grad_tracking(tmpdir, norm_type, log_interval, rtol=5e-3,):
    os.environ['PL_DEV_DEBUG'] = '1'

    # rtol=5e-3 respects the 3 decimals rounding in `.grad_norms` and above

    reset_seed()

    # use a custom grad tracking module and a list logger
    model = ModelWithManualGradTracker(norm_type)

    trainer = Trainer(
        default_root_dir=tmpdir,
        max_epochs=3,
        track_grad_norm=norm_type,
        row_log_interval=log_interval,
    )
    result = trainer.fit(model)

    assert result == 1, "Training failed"
    logged_metrics = trainer.dev_debugger.logged_metrics
    expected_grad_norms = model.stored_grad_norms[::log_interval]
    # Remove first batch and include last batch?
    # 
    
    assert len(logged_metrics) == len(expected_grad_norms)

    # compare the logged metrics against tracked norms on `.backward`
    for mod, log in zip(expected_grad_norms, logged_metrics):
        common = mod.keys() & log.keys()

        log, mod = [log[k] for k in common], [mod[k] for k in common]

        assert np.allclose(log, mod, rtol=rtol)
```

File: tests/models/test_grad_norm.py (hunk 1 above)
Comment (awaelchli) [reply tests/models/test_grad_norm.py:84]: yes, silly of me

**Comment by awaelchli** [comment]
@Tim-Chard Thanks, yes this test you mention is a bit complex and partially re-implements the functionality it is supposed to check. It's better not to add the row_log_interval there.

> You beat me to it!

sorry, it's my "on call" shift so I try to deal with these bug reports as fast as possible because of upcoming release.

**Comment by Tim-Chard** [comment]
> sorry, it's my "on call" shift so I try to deal with these bug reports as fast as possible because of upcoming release.

Don't be, it is good to see the project can squash bugs this quickly.

**Review by rohitgr7 — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by Tim-Chard — APPROVED** [review]

**Review by Tim-Chard — APPROVED** [review]

**Review by Borda — APPROVED** [review]
LGTM 🐰
